### DISCUSSION TITLE
Strategy for ensuring all queries have indexes

### DISCUSSION THREAD

**Discussion by LinusU** [prisma/prisma#15270] [Q&A]
We are currently migrating from MongoDB to MySQL + Prisma. When we were using MongoDB, we had the following strategy for making sure that every query had an index:

- We created index with names that were random strings
- Each query (that didn't include `_id`) always had a `hint` parameter, with the name of the index as the value
- When changing a query we changed/created new index as appropriate

We would like to something similar in MySQL, but there is currently no parameter like `hint` to supply to e.g. `findMany`. This makes it hard to find a good place to put the index name, and it makes it hard to enforce that every query has the name.

I would love some insight into how other organisations are dealing with this!

**Comment by austincrim** [comment]
Hey @LinusU 👋, apologies for the delay in getting to your question.

As far as I know, MySQL will handle indexes for you without any kind of hinting. If you see any unexpected poor performance from some queries, it would be worth investigating your indexing strategy or even opening a [bug report](https://github.com/prisma/prisma/issues/new?assignees=&labels=kind%2Fbug&template=bug_report.yml).

We recently released the tracing preview feature that could help with identifying problem queries.

**Reply by LinusU** [reply]
Hello @austincrim,

I think that you misunderstood my goal a little. I'm trying to prevent any query from not having an index. I'm not really interested in telling my SQL engine which index to use, but in our code base I want to enforce that there is an index created for every query.

Something very awesome would be if the `where` clause to `*Many` could behave the same way as the `*One`, only with any index instead of only unique indexes. This would be more complicated with partial indexes, which I also think is a must have feature 😅 

But I'm also very open to answers that could involve outside tools (like som kind of static checker), or just organisational policies. Which is why I opened this as a discussion ☺️

**Reply by janpio** [reply]
That is an interesting suggestion that I have not heard before. We could indeed _in theory_ enforce index existance via Prisma Client queries at runtime or via types.

Is it good practice though to enforce indexes _everywhere_, all the time?

A way to build this yourself (for the runtime part) could be a middleware. It could parse the `where` and correlated that with the schema file content to decide if the wanted indexes are present.

**Reply by vladi-stevanovic** [reply]
👋 Hello @LinusU did you perhaps have a chance to check Jan's recommendation? Let us know if you still have any questions, or if you've already solved it, we'd love to know how! 

It would be super useful for other users facing this problem! 🙌

**Reply by LinusU** [reply]
I haven't had time to look into this as much as I wanted, but the strategy that we have settled on for now is:

> In order to make sure that our usage of Postgres is performant, every query that doesn't query on `id` most indicate which index is being used in a special comment. If you create a new query, you should also make sure that an existing index covers it, or create a new index. The comment should look like so:
>
> ```typescript
> // [query index: <index name>]
> ```

Then we enforce it in code review. We can also see which indexes are actually being used by grepping for them.

**Reply by pperera** [reply]
@LinusU did you get anywhere with this? I'm hitting a similar issue/desire in my org too!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
